vllm/entrypoints/cli/serve.py

```diff
@@ -26,11 +29,141 @@
+
+
+class APIServerProcessManager:
+    """Manages a group of API server processes.
+    
+    Handles creation, monitoring, and termination of API server worker
+    processes. Also monitors extra processes to check if they are healthy.
+    """
+
+    def __init__(
+        self,
+        listen_address: str,
+        sock: Any,
+        args: argparse.Namespace,
+        num_servers: int,
+        input_addresses: list[str],
+        output_addresses: list[str],
+        stats_update_address: Optional[str] = None,
+        extra_processes_to_healthcheck: Optional[list[
+            multiprocessing.Process]] = None,
+    ):
+        """Initialize and start API server worker processes.
+        
+        Args:
+            listen_address: Address to listen for client connections
+            sock: Socket for client connections
+            args: Command line arguments
+            num_servers: Number of API server processes to start
+            input_addresses: Input addresses for each API server
+            output_addresses: Output addresses for each API server
+            stats_update_address: Optional stats update address
+            extra_processes_to_healthcheck: Optional list of additional 
+                processes to monitor
+        """
+        self.listen_address = listen_address
+        self.sock = sock
+        self.args = args
+
+        # Store additional processes to monitor
+        self.extra_processes_to_healthcheck = (extra_processes_to_healthcheck
+                                               or [])
+
+        # Start API servers
+        spawn_context = multiprocessing.get_context("spawn")
+        self.processes: list[SpawnProcess] = []
+
+        for i, in_addr, out_addr in zip(range(num_servers), input_addresses,
+                                        output_addresses):
+            client_config = {
+                "input_address": in_addr,
+                "output_address": out_addr,
+                "client_index": i
+            }
+            if stats_update_address is not None:
+                client_config["stats_update_address"] = stats_update_address
+
+            proc = spawn_context.Process(target=run_api_server_worker,
+                                         name=f"ApiServer_{i}",
+                                         args=(listen_address, sock, args,
+                                               client_config))
+            self.processes.append(proc)
+            proc.start()
+
+        logger.info("Started %d API server processes", len(self.processes))
+
+        # Casting due to mypy error
+        process_list: list[multiprocessing.Process] = cast(
+            list[multiprocessing.Process], self.processes)
+        # Shutdown only the API server processes on garbage collection
+        # The extra processes are managed by their owners
+        self._finalizer = weakref.finalize(self, shutdown, process_list)
+
+    def wait_for_completion_or_failure(self) -> None:
```

Comment:
Instead of this being a method on this class I think it makes sense to have a separate utility method similar to `wait_for_engine_startup, which takes optional engine manager, api server manager, and coordinator, and then just waits on all the contained procs and calls close on them in a finally block.

Reply:
ok refactored to this utility. Had to update tests as well.